src/stage2/interpreter.jl

```diff
@@ -35,22 +40,161 @@ struct ADInterpreter <: AbstractInterpreter
     # Level 1 == Gradients
     # Level 2 == Seconds Derivatives
     # and so on
-    code::OffsetVector{Dict{MethodInstance, Any}}
+    opt::OffsetVector{Dict{MethodInstance, CodeInstance}}
+    unopt::Union{OffsetVector{Dict{Union{MethodInstance, InferenceResult}, Cthulhu.InferredSource}}, Nothing}
+
     native_interpreter::NativeInterpreter
     current_level::Int
     msgs::Vector{Tuple{Int, MethodInstance, Int, String}}
 end
-change_level(interp::ADInterpreter, new_level::Int) = ADInterpreter(interp.code, interp.native_interpreter, new_level, interp.msgs)
+change_level(interp::ADInterpreter, new_level::Int) = ADInterpreter(interp.opt, interp.unopt, interp.native_interpreter, new_level, interp.msgs)
 raise_level(interp::ADInterpreter) = change_level(interp, interp.current_level + 1)
 lower_level(interp::ADInterpreter) = change_level(interp, interp.current_level - 1)
 
+Cthulhu.get_optimized_code(interp::ADInterpreter, curs::ADCursor) = interp.opt[curs.level][curs.mi]
+Cthulhu.AbstractCursor(interp::ADInterpreter, mi::MethodInstance) = ADCursor(0, mi)
+
+# This is a lie, but let's clean this up later
+Cthulhu.can_descend(interp::ADInterpreter, mi::MethodInstance, optimize::Bool) = true
+
+function Cthulhu.lookup(interp::ADInterpreter, curs::ADCursor, optimize::Bool; allow_no_src::Bool=false)
```

Comment:
```suggestion
function Cthulhu.lookup(interp::ADInterpreter, curs::ADCursor, optimize::Bool)
```
No need to overload the kwarg method (it's only used in the Cthulhu test suite).